### PR TITLE
Best solutions

### PR DESCRIPTION
**Pull request by krulis-martin**

### CODE REVIEW

File: app/model/repository/AssignmentSolutions.php [hunk 1]
```diff
@@ -86,44 +87,90 @@ private function findValidSolutionsForAssignments(array $assignments, User $user
...
+      return $best;
+    }
+
+    if ($solution->isAccepted()) {
+      return $solution;
+    }
+
+    if ($best->getTotalPoints() < $solution->getTotalPoints()
+      || ($best->getTotalPoints() === $solution->getTotalPoints()
+      && $best->getSolution()->getCreatedAt() < $solution->getSolution()->getCreatedAt())) {
```
Comment (Neloop) [inline app/model/repository/AssignmentSolutions.php]: This line is unnecessary, solutions are ordered in database, either remove this line, or ordering in database...

File: app/model/view/AssignmentSolutionViewFactory.php [hunk 2]
```diff
@@ -33,21 +40,40 @@ class AssignmentSolutionViewFactory {
...
+      $bestId = reset($best)->getId();
+      return $bestId === $solution->getId();
+    }
+    throw new InternalServerException("Something is terribly wrong. We have a solution, yet we cannot find any solutions for corresponding assignment and author.");
+  }
 
   /**
    * Parametrized view.
    * @param AssignmentSolution $solution
+   * @param bool|array|null $bestSolutionsHints If bool value is provided, it holds the `isBestSolution` value already.
```
Comment (Neloop) [inline app/model/view/AssignmentSolutionViewFactory.php]: We gotta love PHP!

File: app/model/repository/AssignmentSolutions.php [hunk 3]
```diff
@@ -102,28 +182,39 @@ public function findBestSolutionsForAssignments(array $assignments, User $user):
...
+   */
+  public function filterBestSolutions(array $solutions) {
+    $result = [];
+    $validSolutions = self::filterValidSolutions($solutions);
+    foreach ($validSolutions as $solution) {
+      $assignment = $solution->getAssignment();
+      if ($assignment === null) {
         continue;
       }
+      $assignmentId = $assignment->getId();
```
Comment (Neloop) [inline app/model/repository/AssignmentSolutions.php]: Unused variable

File: app/model/repository/AssignmentSolutions.php [hunk 4]
```diff
@@ -102,28 +182,39 @@ public function findBestSolutionsForAssignments(array $assignments, User $user):
...
 
-      if ($solution->isAccepted()) {
-        $result[$assignment->getId()] = $solution;
+      $author = $solution->getSolution()->getAuthor();
+      if ($author === null) {
         continue;
       }
+      $key = $assignment->getId() . ':' . $author->getId();
```
Comment (Neloop) [inline app/model/repository/AssignmentSolutions.php]: You almost got me there :-) ... I almost wanted to point out that you do not have nested arrays...

Comment (krulis-martin) [reply app/model/repository/AssignmentSolutions.php]: OK, I'll add a comment

File: app/model/view/AssignmentSolutionViewFactory.php (hunk 2 above)
Comment (krulis-martin) [reply app/model/view/AssignmentSolutionViewFactory.php]: Yesss, we do! :D

**Pull request merged**